### ISSUE TITLE
New interface for colormap selection

### ISSUE THREAD

**Issue by Czaki** [napari/napari#1779]
## 🚀 Feature

As was discussed at the napari euro dev meeting I would like to start a discussion about adding an interface to control parameters of presentation. 

In the current version of napari, there is a need for a manual set colormap of every layer. This is annoying when having multiple channels (layer) data. 

Napari should provide some interface that allows simpler adjust parameters of data presentation. 

Current:

![napari-set-colormap](https://user-images.githubusercontent.com/3826210/97192763-86e48580-17a8-11eb-83f9-4ac848637982.gif)

And after load the next datapoint all steps need to be done again. 

The next problem is that the user needs to know which colormap is hidden under the given name. 
![napari-set-colormap2](https://user-images.githubusercontent.com/3826210/97192963-c0b58c00-17a8-11eb-901a-0f98ab2aa83b.gif)

In my tool (PartSeg) I use a combo box that shows colormaps instead of its name. 

![partseg-set-colormap](https://user-images.githubusercontent.com/3826210/97185928-f8b8d100-17a0-11eb-9c2a-c6f3c55c576a.gif)

In my tool, I use a bar that allows easy colormap pear each channel, but I'm not sure if it is a good approach for napari. 

![Zrzut ekranu z 2020-10-26 15-49-36](https://user-images.githubusercontent.com/3826210/97187620-e9d31e00-17a2-11eb-95b8-23ee27388a0a.png)

But the list of selected colormaps is not reset when loading a new image. Also when loading an image with lower number of channels. 

![partseg-load-image](https://user-images.githubusercontent.com/3826210/97189710-2d2e8c00-17a5-11eb-9f87-cfcce6a65422.gif)

Maybe the best option is to allow save the current set of visualization parameters to the same profile and allow user to select this profile from some dropdown list? When #1089 is implemented maybe there should be some view control for the whole group. 

As napari has saved state using YAML file it is possible to add colormap creator which allows a user to create better colormap for his project. Here Is colormap creator from my tool (with show control of the list of selectable colormaps):

![partseg-colormap](https://user-images.githubusercontent.com/3826210/97191933-aaf39700-17a7-11eb-9ab1-1f82b708a099.gif)

Also, color bar which inform about mapping colormap to pixel brightness could be nice feature. Here one from my tool: 
![partseg-colorbar](https://user-images.githubusercontent.com/3826210/97193537-68cb5500-17a9-11eb-92c7-58a9fb4a0207.gif)

I creat this to show few ideas which could improve napari interface for users who does not use API.

**Comment by jni** [comment]
Thanks @Czaki! Actually we had an independent discussion about this at the US/Pacific meeting, so this is timely, and clearly an area of need for napari!

My proposal was to have a panel where you could have all the colormaps on one side, all the layers on the other, and you could either drag layers to colormaps to make them visible under that colormap, or drag them to an "invisible" section. I think that's somewhat similar to what you're doing with your color bar?

I would prefer we did *not* offer colormap creation. Creating a good colormap is hard work, and I'd like to avoid napari screenshots appearing with a bunch of terrible colormaps.

**Comment by Czaki** [comment]
> My proposal was to have a panel where you could have all the colormaps on one side, all the layers on the other, and you could either drag layers to colormaps to make them visible under that colormap, or drag them to an "invisible" section. I think that's somewhat similar to what you're doing with your color bar?

I do not understand? If this is related to the mechanism of selection which colormap is used? In my case, the change of colormap presented on colorbar is caused by a change of the active layer. 

> I would prefer we did _not_ offer colormap creation. Creating a good colormap is hard work, and I'd like to avoid napari screenshots appearing with a bunch of terrible colormaps.

Maybe. I do not have enough knowledge and resources to create a complete set of colormaps both for screenshot and provide clear readability for set parameters of segmentation algorithms.

**Comment by GenevieveBuckley** [comment]
We discussed this at the US/Australia dev meeting today, here are my thoughts:
* I am +1 on the idea of displaying a little picture of the colormaps next to their names in the colormap drop down menu
* I am +1 on the idea of having a colorbar that can be displayed (a bit like the scale bar, I think this will be useful for making figures for publication)
* I am against the idea of supporting user creation of colormaps in the way you suggest above. I don't like this idea for two reasons:
    * We want napari to be used primarily for scientific visualization. This feature raises the possibility of users creating misleading/unintuitive colormaps (either deliberately or unwittingly). This seems like an obvious pitfall we want to discourage.
    * I don't think the advantage outweighs the maintenance burden in this case. I do not think this would be a good use of developer & maintainer time.

**Comment by Czaki** [comment]
> I am against the idea of supporting user creation of colormaps in the way you suggest above. I don't like this idea for two reasons:
> 
>    * We want napari to be used primarily for scientific visualization. This feature raises the possibility of users creating misleading/unintuitive colormaps (either deliberately or unwittingly). This seems like an obvious pitfall we want to discourage.
> 
>    * I don't think the advantage outweighs the maintenance burden in this case. I do not think this would be a good use of developer & maintainer time.

In such a case, I suggest include in napari more colormaps.

**Comment by jni** [comment]
@Czaki we've actually had some good discussions about this, but we haven't figured out a good way to implement it. See #268 (a while ago!!! :joy:), and references within it. What we want is a *short* curated list that can be expanded when the user types in it, much like the license chooser when creating a new repo on GitHub. If you have ideas with how to implement this in Qt, that would be amazing! I can even envision something like typing `#fe00a1` to get a linear colormap to that particular color.

**Comment by GenevieveBuckley** [comment]
> In such a case, I suggest include in napari more colormaps.

This could be a good solution - which specific colourmaps do you think would be most useful and are missing currently?

**Comment by Czaki** [comment]
@jni using Completer from Qt searchable combo box is really simple (see sample implementation https://github.com/4DNucleome/PartSeg/blob/master/package/PartSeg/common_gui/searchable_combo_box.py).

@GenevieveBuckley I'm not sure. My first approach is to add all colormaps from matplotlib, but #1183 is not merged yet. So there is no option to add preselection (see below) of ones visible in dropdown may be hard. 

![Zrzut ekranu z 2020-10-31 10-00-00](https://user-images.githubusercontent.com/3826210/97775301-df78a180-1b5f-11eb-8912-84002e6fe3c9.png)

**Comment by jni** [comment]
> using Completer from Qt searchable combo box is really simple (see sample implementation

This is great! It's not quite what we want though, I think (just looking at the code, correct me if I'm wrong): you still have the full list in the dropdown. What we really want is to show *simplified* short list, of our specified colormaps, and then if you start typing then it autocompletes from a much bigger list. Does that make sense? Do you think that would be easy?

However, we could have a compromise by sorting the list according to most common. Not sure. But we originally started with all matplotlib colormaps and people found it too overwhelming.

**Comment by Czaki** [comment]
> This is great! It's not quite what we want though, I think (just looking at the code, correct me if I'm wrong): you still have the full list in the dropdown. What we really want is to show _simplified_ short list, of our specified colormaps, and then if you start typing then it autocompletes from a much bigger list. Does that make sense? Do you think that would be easy?

I think yes. The first idea is to check if the user-provided some input and if Yes then add additional items (self call of addItems). And after selection is made then remove redundant items.

**Comment by ptlzon** [comment]
its been 5 years, any updates?

**Comment by jni** [comment]
For linear colormaps, there is #7600 (last backlink immediately above your comment), which will be available in the upcoming release 0.6.0. We will not support creating segmented colormaps directly in napari. We already support all matplotlib colormaps from the API. (e.g. `image_layer.colormap = 'RdBu'`) If you need it from the UI, I *think* pyapp-kit/superqt#278 might finally solve this. Since #7555 is tracking this, I think we can close this issue. @psobolewskiPhD, thoughts?

**Comment by ptlzon** [comment]
Thanks for the response. I was looking for placing the colorbar legend. However, it seems that some code can draw it. But I think it is preferable to have a GUI to show/hide the colorbar legend.
For future reader, here is my example to draw the already selected colormap and map to the values (for my case it's [0,255].

`
# Create a clean colorbar with fixed ticks at 0, 50, 100, 150, 200, 250
import matplotlib.pyplot as plt
from matplotlib.figure import Figure
from matplotlib.backends.backend_qt5agg import FigureCanvasQTAgg
from qtpy.QtWidgets import QWidget, QVBoxLayout
import numpy as np

# Get the current viewer and active layer
viewer = napari.current_viewer()
layer = viewer.layers.selection.active  # Get the currently selected layer

# Create a widget to hold the colorbar
colorbar_widget = QWidget()
layout = QVBoxLayout()
colorbar_widget.setLayout(layout)

# Create a matplotlib figure for the colorbar
fig = Figure(figsize=(6, 1.2))
canvas = FigureCanvasQTAgg(fig)
ax = fig.add_subplot(111)

# Get the colormap from the layer
if hasattr(layer, 'colormap'):
    cmap_name = layer.colormap.name
    cmap = plt.get_cmap(cmap_name)
else:
    # Default colormap if none is found
    cmap = plt.get_cmap('viridis')

# Get the contrast limits (min/max values for display)
if hasattr(layer, 'contrast_limits'):
    vmin, vmax = layer.contrast_limits
else:
    vmin = np.min(layer.data)
    vmax = np.max(layer.data)

# Create a colorbar with the layer's colormap and data range
norm = plt.Normalize(vmin, vmax)
sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)
sm.set_array([])

# Hide the main axes completely
ax.set_visible(False)

# Create the colorbar
cbar = fig.colorbar(sm, ax=ax, orientation='horizontal')

# Set the custom label
cbar.set_label("CNT Volume (Relative Scale)")

# Explicitly set ticks at 0, 50, 100, 150, 200, 250
# Only show ticks that are within the data range
fixed_ticks = np.array([0, 50, 100, 150, 200])
tick_values = fixed_ticks[(fixed_ticks >= vmin) & (fixed_ticks <= vmax)]

# If no fixed ticks are in range, fall back to linspace
if len(tick_values) == 0:
    tick_values = np.linspace(vmin, vmax, 5)

# Always include vmin and vmax if they're not already in the ticks
if vmin not in tick_values:
    tick_values = np.insert(tick_values, 0, vmin)
if vmax not in tick_values and vmax > tick_values[-1]:
    tick_values = np.append(tick_values, vmax)

cbar.set_ticks(tick_values)
cbar.set_ticklabels([f'{int(v)}' if v.is_integer() else f'{v:.1f}' for v in tick_values])

# Adjust layout to make sure everything fits
fig.tight_layout()

# Add the colorbar to the widget
layout.addWidget(canvas)

# Add the widget to the viewer
viewer.window.add_dock_widget(colorbar_widget, name='CNT Volume Colorbar', area='bottom')
`

**Comment by jni** [comment]
Wow. I don't know how that issue went all that time without any core dev response. 😲 Sorry about that!

I'll close this one, and probably reopen that one. Indeed it remains on our to-do. We've gone through a refactor of our overlay machinery but it's still too difficult to implement such a simple thing. 😕

**Comment by tlambert03** [comment]
@jni, do you see https://github.com/pyapp-kit/superqt/pull/278 as mandatory for you to be able to use cmap?  It looks more like a subtle UI change, rather than a fundamentally different presentation.  The existing `QColormapComboBox` was really kinda designed as a drop-in for napari. (https://github.com/pyapp-kit/superqt/pull/278 seems to just save a click)

**Comment by psobolewskiPhD** [comment]
@tlambert03 eep, still haven't gotten to that one 👀 
Maybe I can find some time today.
Anyhow, I don't think that's a requirement at all, but more like sugar to make the medicine go down? 
I have an issue open with thoughts using cmap:
https://github.com/napari/napari/issues/7555

**Comment by tlambert03** [comment]
Yeah no worries, there's no hurry there.  Just didn't see why @jni would be suggesting that that PR specifically solves this one

**Comment by jni** [comment]
Probably I misunderstood what it does. What I'm after is described in #268: if you click, you get a short, curated list of "recommended" colormaps, but if you start typing, you filter from a much bigger list — in this case, the union of mpl, vispy, and napari colormaps. (And maybe cmap?)

**Comment by tlambert03** [comment]
yep, that's the intention of `QColormapComboBox`... starts with a short list of recommended.  And additional ones can be added, not  by typing, but by selecting the "add colormap..." option at the bottom of the list, which brings up the full catalog, filterable by colormap type, and you can type to filter.

https://github.com/user-attachments/assets/eeb4f253-8236-492e-83f9-fe4f4e1eac79

what https://github.com/pyapp-kit/superqt/pull/278 adds is that it it lets you start typing and it will show filtered options from the already "curated" list (not the entire list)...  It does look like if you type the name of a missing item *exactly*, it will indeed use it, but it doesn't give you suggestions as you're typing, and it also doesn't add it to the list (maybe that's the ultimate intention though?  @psobolewskiPhD?)

**Comment by jni** [comment]
Cool! Thanks for the screenshare. Well, I'd be in favour of filtering from the big list. What do you think about that @tlambert03?

> what https://github.com/pyapp-kit/superqt/pull/278 adds is that it it lets you start typing and it will show filtered options from the already "curated" list

We have this already in napari main so I guess that's important for us to start using the superqt widget directly? But it would be a game changer imho to filter from the full list. (While placing the curated elements at the start of the filtered list so that we don't have a regression in functionality.)

**Comment by psobolewskiPhD** [comment]
My aim in https://github.com/pyapp-kit/superqt/pull/278 was, as I tried to explain in https://github.com/napari/napari/issues/7555
to tame the napari colormap list. It's curated, but it's long and alphabetical is annoying. And there is no way to add colormaps from the GUI. So the idea was to have a list that can be quickly filtered and allow users to add from the cmap catalog -- which is freaking awesome.
I didn't want to use the cmap catalog widget, because I like the idea of a set of curated colormaps. We don't need the cognitive load on a new user of *having* to filter or whatever.
But I do want it to be easier to add new colormaps.

**Comment by psobolewskiPhD** [comment]
I think our posts crossed paths.
As i noted in my post above, I don't like the idea of having every colormap known to man in a list out of the box.
It's overwhelming. I prefer the idea of a list that can be filtered and added to from a catalog.

**Comment by tlambert03** [comment]
> I didn't want to use the cmap catalog widget, because I like the idea of a set of curated colormaps.

I'm a bit confused... the `QColormapComboBox` *does* start with a set of curated colormaps... you have to click on "add colormap..." to get the comprehensive list where you can filter based on typing.  is your goal to make that second list also a curated (but longer) set of colormaps?  That would definitely be possible too... but feels like something different from https://github.com/pyapp-kit/superqt/pull/278

**Comment by psobolewskiPhD** [comment]
I was replying to @jni who wanted all of the colormaps.
So I like QColormapComboBox, just wanted to be able to filter.

**Comment by tlambert03** [comment]
gotcha 👍.  fwiw you can start typing to select once you click on the combobox.  but if you also want it to visually narrow (rather than jump straight to what you're typing) then indeed, we'll need to keep going with 278

**Comment by psobolewskiPhD** [comment]
@tlambert03 I prefer the filtering because it's more intuitive with multiple colormaps with similar names. But certainly the existing widget is very nice and an upgrade on what we are using currently.
